drivers/modem/quectel-bg9x.c

```diff
@@ -1029,8 +1029,29 @@ static int modem_setup(void)
 			     NULL, 0U, "AT+QIACT=1", &mdata.sem_response,
 			     MDM_CMD_TIMEOUT);
 
-	/* Retry or Possibly Exit. */
-	if (ret < 0 && init_retry_count++ < MDM_INIT_RETRY_COUNT) {
+	/* If there is trouble opening the PDP context, we try to deactivate/reactive it 3 times. */
+	while (ret == -EIO && pdp_act_retry_count < MDM_PDP_ACT_RETRY_COUNT) {
+
+		modem_cmd_send(&mctx.iface, &mctx.cmd_handler,
+			     NULL, 0U, "AT+QIDEACT=1", &mdata.sem_response,
+			     MDM_CMD_TIMEOUT);
+
+		/* If there's an error for the AT+QIDEACT, restart the module. */
+		if (ret < 0) {
+			goto restart;
+		}
+
+		ret = modem_cmd_send(&mctx.iface, &mctx.cmd_handler,
+			     NULL, 0U, "AT+QIACT=1", &mdata.sem_response,
+			     MDM_CMD_TIMEOUT);
+
+		pdp_act_retry_count++;
+	}
+
+	/* Retry or Possibly Exit if there's 3 consecutive failures. */
+	if ((ret < 0 && init_retry_count++ < MDM_INIT_RETRY_COUNT) ||
+		(pdp_act_retry_count >= MDM_PDP_ACT_RETRY_COUNT)) {
+		pdp_act_retry_count = 0;
 		goto restart;
```

Comment:
There might be case when `ret == 0`, but `pdp_act_retry_count == 3`. This happens when `AT+QIACT=1` in the last loop has succeeded. If it succeeded, then you don't want to restart modem. So if you will create a `bg9x_pdp_context_open()` helper function, then you will no longer have to think about `pdp_act_retry_count` at this level.

Also, do not write specifics in comments like "how many times it will fail" before doing something ("... 3 consecutive failures"). MDM_INIT_RETRY_COUNT is actually 10, while MDM_PDP_ACT_RETRY_COUNT is 3. I doubt that those will be in sync for very long time. Actually, "3 consecutive failures." suggests that MDM_INIT_RETRY_COUNT is 3, but it is not. Let C describe the details and just write high level comments describing the general logic behind. Or create helper functions with good names, which hopefull will describe themselves.

Reply:
Fixed.